rfcs/999-kustomize.adoc

```diff
@@ -0,0 +1,197 @@
+it's not appropriate to have a scripting engine running arbitrary code inside a
+server side component such as Shipper, as it would be relatively easy for a
+misbehaved chart to pin down Shipper workers, potentially denying service to
+other well behaved applications.
+
+Second, Helm generates Kubernetes objects by templating YAML files using Go
+templates. Generating well formed YAML files this way is very error prone, as
+one needs to pay attention to correctly balanced quotes, keeping track of
+indentation levels and others. Helm's choice of using Go templates also
+prevents us from instrospecting charts with languages other than Go.
+
+Lastly, Helm charts only allow configuration of values that the chart authors
+have thought of, and wrote templates for. If the need arises to change a value
+that's hardcoded in the chart, or add a completely new one that's not present,
+it's necessary to fork the chart, and add the changes there. We would prefer to
+allow users to apply arbitrary patches to objects instead, as it would reduce,
+or completely remove, the need to fork charts.
+
+link:https://kustomize.io/[Kustomize] brands itself as a Kubernetes native
+configuration management that does not rely on templates to customize
+application configuration. It uses plain YAML artifacts, allowing easier
+instrospection from any programming language. It also allows the application of
+arbitrary patches to any of the resources it manages, allowing more freedom to
+users of off-the-shelf configuration bundles to customize to their heart's
+content.
+
+As Kustomize seems to address the challenges with our current usage of Helm,
+this RFC aims to discuss how a potential implementation of Kustomize would work
+in Shipper.
+
+== Reference level explanation
+
+A common use case for us is to have a Shipper application referring to a
+centrally supported bundle of configuration, with sane defaults for basic
+infrastructure, and a placeholder for the application's Docker image.
+Additionally, applications should be able to enable or disable bigger chunks of
+configuration, such as which optional sidecars (out of either a collection of
+centrally supported sidecars, or completely bespoke ones) should be present.
+Today, that's supported by a Helm chart specified by `.spec.template.chart`,
+and tweaks to the configuration are injected as values through
+`.spec.template.values`, both in the Application object.
+
+Using Kustomize, the Application object would lose both `.spec.template.chart`
+and `.spec.template.values`, having those replaced by a single
+`.spec.template.kustomization`, as exemplified below, containing an inline
+Kustomization object.
+
+.application.yaml
+[source,yaml]
+----
+apiVersion: shipper.booking.com/v1alpha1
+kind: Application
+metadata:
+  name: bikerental
+spec:
+  template:
+    clusterRequirements:
+      regions:
+      - name: eu-nl
+      capabilities: []
+    kustomization:
+      image:
```

Comment:
How does Kustomize know which object in the bundle to apply this patch to?

Reply:
First of all, my mistake, I got the schema wrong ^_^ It should be `images`. Second of all, it's not a patch, it's a transformer, making it somewhat involved from kustomize's side. [Documentation here](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/fields.md#images).